xapian-core/common/parseint.h

```diff
@@ -37,4 +38,21 @@ bool parse_unsigned(const char* p, T& res)
     return true;
 }
 
+template<typename T>
+bool parse_signed(const char* p, T& res)
+{
+    typedef typename std::make_unsigned<T>::type unsigned_type;
+    unsigned_type temp = 0;
+    if (*p == '-' && parse_unsigned(++p, temp) &&
+	temp <= unsigned_type(std::numeric_limits<T>::min())) {
```

Comment:
Here `T` is a signed type so `std::numeric_limits<T>::min()` is negative.  Considering the case when T is `signed char`, `min()` gives -128, and the cast to unsigned_type should give the twos' complement bit pattern for -128 which is `0x80`, and that's 128 as an unsigned value.

So it looks like the test is correct (as things will work similarly for other signed types), but it relies on the fact that casting the min signed value to unsigned gives us its absolute value which seems a little non-obvious (I initially thought there must be a bug here when I spotted that a negative value was getting cast).

Not sure I see a better way to write this though - something like `temp <= unsigned_type(T(-1) - std::numeric_limits<T>::min()) + 1` would work I think, but doesn't seem any clearer.

Perhaps worth noting in a comment what `unsigned_type(std::numeric_limits<T>::min())` gives us though.

Reply:
fixed in a fixup commit